AGILE/Commands.cs

```diff
@@ -1639,7 +1639,8 @@ private int ExecuteAction(Action action, ref byte newRoom, ref bool exit)
                             ShowPicture(false);
                             newRoom = state.CurrentRoom = state.Vars[Defines.CURROOM];
                             textGraphics.UpdateStatusLine();
-                            exit = true;
+                            // return 0 so we can start executing the beginning of Logic 0
```

Comment:
This bit I'm not 100% sure about. Are we sure that it starts executing the current Logic (which in this case is Logic 0) from the start again? Without leaving the ExecuteLogic method?

Stepping through the code in my mind: We're executing Logic 0 and restore.game is being run. This is part way through Logic 0. After the restore game code is executed, it returns 0, which would force the ExecuteLogic method to start executing the current Logic (in this case Logic 0) from the start again. It is in effect a "goto" that goes back to the top, i.e. an unconditional branch to 0 (equivalent to nextActionNum = 0). Do you agree?

If so, then it means that it doesn't leave ExecuteLogic, since exit=true has been removed, so would not run the new code within the "while (executeLogicLoop)" in the Interpreter class until it exits Logic 0 for some other reason, e.g. getting to the end of Logic 0, or a new room.

I keep thinking that maybe the way I've got AGILE using the return value of Newroom to decide whether to repeat the outer while loop might be the issue. Currently changing room is the only way it decides to run that loop again. It seems like this might not have been true for the original AGI interpreter.

I try not to peek into the NAGI source code if I can help it, but given the absence of the original AGI interpreter source code in this case, then I think NAGI is the next best source of information, being based on disassembled code. NAGI's equivalent of the outer while loop is enlightening, as it has additional comments that the original AGI interpreter source does not have:

		while (logic_call(0) == 0)	// logic 0
		{
			// comes here if we need to restart logic0
			// newroom, restart, restore
			state.var[V09_BADWORD] = 0;
			state.var[V05_OBJBORDER] = 0;
			state.var[V04_OBJECT] = 0;
			flag_reset(F02_PLAYERCMD);	// player has not issued command line
			old_score = state.var[V03_SCORE];
		} 

The code is identical to the original AGI interpreter source, but it has those two comments at the top that tells us a bit more info. So apparently this while loop should repeat not only in the newroom scenario, but also restart and restore. So in the case of the original AGI interpreter, NULL is defined as 0, so it likewise will loop again if _CallLogic returns 0, which means that _CallLogic must return 0/NULL for newroom, restore and restart.

I _think_ that the issue with AGILE might be that it uses only newroom to decide whether to loop again, and this  links back to my earlier comment that I was probably trying to be too clever to have the NewRoom call wrapping the ExecuteLogic call. Currently ExecuteLogic returns the new room number, if it has changed, otherwise the current room number. The original AGI interpreter didn't do that but instead returned a value of 0 in the case of newroom, restore or restart, otherwise the logic offset value (which it doesn't appear to have used for anything).

Let me think about how ExecuteLogic and NewRoom can be refactored to work more like the original AGI interpreter. I agree that we need most of that body of the while loop that you have below, but NewRoom might need to be rethought. It might be that refactoring ExecuteLogic and NewRoom to work more like the original AGI interpreter will need that your "return 0" change might not be required, as that would mean that it would loop back in the while loop and execute Logic 0 again from the start, and will also have executed the body of the while loop (i.e. the new bits you added below).

Hope that all makes sense. I've been typing for the past hour without realising this comment has grown so big!